File: NETGEAR_Genie_Win8/NETGEAR_Genie_Win8/ViewModels/NetworkMapModel.cpp
#include "pch.h"
#include "ViewModels\NetworkMapModel.h"

using namespace NETGEAR_Genie_Win8::Data;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::ApplicationModel::Resources::Core;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;
using namespace Windows::UI::Xaml::Media::Imaging;

//
// DeviceCommon
//

DeviceCommon::DeviceCommon(Platform::String^ uniqueId)
{
	_uniqueId = uniqueId;
	
}

String^ DeviceCommon::UniqueId::get()
{
	return _uniqueId;
}

void DeviceCommon::UniqueId::set(String^ value)
{
	if (_uniqueId != value)
	{
		_uniqueId = value;
		OnPropertyChanged("UniqueId");
	}
}

//
// DeviceItem
//

DeviceItem::DeviceItem(Platform::String^ uniqueId, Platform::String^ deviceName, Platform::String^ IPaddress, Platform::String^ signalStrength, Platform::String^ linkRate, Platform::String^ MACaddress, DeviceGroup^ group)
							   : DeviceCommon(uniqueId)
{
	_deviceName = deviceName;
	_IPaddress = IPaddress;
	_signalStrength = signalStrength;
	_linkRate = linkRate;
	_MACaddress = MACaddress;
	_group = group;
}

String^ DeviceItem::DeviceName::get()
{
	return _deviceName;
}

void DeviceItem::DeviceName::set(String^ value)
{
	if (_deviceName != value)
	{
		_deviceName = value;
		OnPropertyChanged("DeviceName");
	}
}

String^ DeviceItem::IPAddress::get()
{
	return _IPaddress;
}

void DeviceItem::IPAddress::set(String^ value)
{
	if (_IPaddress != value)
	{
		_IPaddress = value;
		OnPropertyChanged("IPAddress");
	}
}

String^ DeviceItem::SignalStrength::get()
{
	return _signalStrength;
}

void DeviceItem::SignalStrength::set(String^ value)
{
	if (_signalStrength != value)
	{
		_signalStrength = value;
		OnPropertyChanged("SignalStrength");
	}
}

String^ DeviceItem::LinkRate::get()
{
	return _linkRate;
}

void DeviceItem::LinkRate::set(String^ value)
{
	if (_linkRate != value)
	{
		_linkRate = value;
		OnPropertyChanged("LinkRate");
	}
}

String^ DeviceItem::MACAddress::get()
{
	return _MACaddress;
}

void DeviceItem::MACAddress::set(String^ value)
{
	if (_MACaddress != value)
	{
		_MACaddress = value;
		OnPropertyChanged("MACAddress");
	}
}

DeviceGroup^ DeviceItem::Group::get()
{
	return _group.Resolve<DeviceGroup>();
}

void DeviceItem::Group::set(DeviceGroup^ value)
{
	if (Group != value)
	{
		_group = value;
		OnPropertyChanged("Group");
	}
}

//
// DeviceGroup
//

DeviceGroup::DeviceGroup(Platform::String^ uniqueId)
	: DeviceCommon(uniqueId)
{
	_items = ref new Vector<DeviceItem^>();
}

IObservableVector<DeviceItem^>^ DeviceGroup::Items::get()
{
	return _items;
}

//
// DeviceSource
//

DeviceSource::DeviceSource()
{
	_deviceGroups = ref new Vector<DeviceGroup^>();
	
	auto group = ref new DeviceGroup("DeviceGroup");
	group->Items->Append(ref new DeviceItem("Router",
		"WNR3500Lv2",
		"192.168.1.1",
		"",
		"",
		"20:4E:7F:04:31:3C",
		group));
	group->Items->Append(ref new DeviceItem("LocalDevice",
		"android-25531554966beee3",
		"192.168.1.25",
		"78%",
		"5.5Mbps",
		"D4:20:6D:D6:37:D6",
		group));
	_deviceGroups->Append(group);
}

IObservableVector<DeviceGroup^>^ DeviceSource::DeviceGroups::get()
{
	return _deviceGroups;
}

static DeviceSource^ _deviceSource = nullptr;

void DeviceSource::Init()
{
	if (_deviceSource == nullptr)
	{
		_deviceSource = ref new DeviceSource();
	}
}

IIterable<DeviceGroup^>^ DeviceSource::GetGroups(String^ uniqueId)
{
	Init();
	return _deviceSource->DeviceGroups;
}

File: GenieWin8/GenieWin8/ViewModels/NetworkMapModel.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using Windows.ApplicationModel.Resources.Core;
using Windows.Foundation;
using Windows.Foundation.Collections;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Media.Imaging;
using System.Collections.Specialized;
using GenieWin8.DataModel;
using Windows.Security.ExchangeActiveSyncProvisioning;

namespace GenieWin8.Data
{  
    public class DeviceGroup
    {
        public DeviceGroup(Node node)
        {
            this._node = node;
        }

        private Node _node = null;
        public Node NODE
        {
            get { return this._node; }
        }
    }

    public sealed class DeviceSource
    {
        //private static DeviceSource _deviceSource = new DeviceSource();

        private ObservableCollection<DeviceGroup> _deviceGroups = new ObservableCollection<DeviceGroup>();
        public ObservableCollection<DeviceGroup> DeviceGroups
        {
            get { return this._deviceGroups; }
        }

        public static IEnumerable<DeviceGroup> GetGroups()
        {
            DeviceSource _deviceSource = new DeviceSource();
            return _deviceSource.DeviceGroups;
        }

        DeviceGroup group;
        public static DeviceGroup GetGroup(string uniqueId)
        {
            // 对于小型数据集可接受简单线性搜索
            DeviceSource _deviceSource = new DeviceSource();
            var matches = _deviceSource.DeviceGroups.Where((group) => group.NODE.uniqueId.Equals(uniqueId));
            if (matches.Count() == 1) return matches.First();
            return null;
        }

        public DeviceSource()
        {
            //Dictionary<string, Dictionary<string, string>> attachDeviceAll = new Dictionary<string, Dictionary<string, string>>();
            List<List<string>> attachDeviceAll = new List<List<string>>();
            attachDeviceAll = NetworkMapInfo.attachDeviceDic;
            UtilityTool util = new UtilityTool();
            string loacalIp = util.GetLocalHostIp();

            Node NodeRouter = new Node();
            NodeRouter.uniqueId = "Router";
            NodeRouter.deviceName = MainPageInfo.model;
            NodeRouter.RouterFirmware = MainPageInfo.firmware;
            NodeRouter.IPaddress = NetworkMapInfo.geteway;
            NodeRouter.MACaddress = WifiInfoModel.macAddr;
            var routerGroup = new DeviceGroup(NodeRouter);
            this.DeviceGroups.Add(routerGroup);

            if (attachDeviceAll.Count == 0)
            {
                Node NodeDevice = new Node();
                NodeDevice.uniqueId = "LocalDevice";
                EasClientDeviceInformation easClientDeviceInformation = new EasClientDeviceInformation();
                NodeDevice.deviceName = easClientDeviceInformation.FriendlyName;
                NodeDevice.IPaddress = loacalIp;
                NodeDevice.MACaddress = "";
                NodeDevice.connectType = "wireless";
                NodeDevice.AccessControl = "";
                var group = new DeviceGroup(NodeDevice);
                this.DeviceGroups.Add(group);
            } 
            else
            {
                //foreach (string key in attachDeviceAll.Keys)                                                //先找出本设备放在设备列表第二个（第一个位路由器）
                foreach (List<string> deviceInfo in attachDeviceAll)
                {
                    Node NodeDevice = new Node();

                    //if (loacalIp == attachDeviceAll[key]["Ip"])
                    if (loacalIp == deviceInfo.ElementAt(0))
                    {
                        NodeDevice.uniqueId = "LocalDevice";
                        bool bFound = false;
                        if (NetworkMapInfo.fileContent != "")
                        {
                            string[] AllDeviceInfo = NetworkMapInfo.fileContent.Split(';');
                            for (int i = 0; i < AllDeviceInfo.Length; i++)
                            {
                                if (AllDeviceInfo[i] != "" && AllDeviceInfo[i] != null)
                                {
                                    string[] DeviceInfo = AllDeviceInfo[i].Split(',');
                                    //if (DeviceInfo[0] == key)
                                    if (DeviceInfo[0] == deviceInfo.ElementAt(2))
                                    {
                                        bFound = true;
                                        if (DeviceInfo[1] != "")
                                            NodeDevice.deviceName = DeviceInfo[1];
                                        else
                                            //NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                                            NodeDevice.deviceName = deviceInfo.ElementAt(1);

                                        if (DeviceInfo[2] != "")
                                            NodeDevice.deviceType = DeviceInfo[2];
                                        else
                                            NodeDevice.deviceType = "networkdev";
                                    }
                                }
                            }
                            if (!bFound)
                            {
                                //NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                                NodeDevice.deviceName = deviceInfo.ElementAt(1);
                                NodeDevice.deviceType = "networkdev";
                            }
                        }
                        else
                        {
                            //NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                            NodeDevice.deviceName = deviceInfo.ElementAt(1);
                            NodeDevice.deviceType = "networkdev";
                        }
                        //NodeDevice.IPaddress = attachDeviceAll[key]["Ip"];
                        //if (attachDeviceAll[key].ContainsKey("LinkSpeed"))
                        //{
                        //    NodeDevice.linkRate = attachDeviceAll[key]["LinkSpeed"] + "Mbps";
                        //}
                        //else
                        //{
                        //    NodeDevice.linkRate = "Mbps";
                        //}

                        //if (attachDeviceAll[key].ContainsKey("Signal"))
                        //{
                        //    NodeDevice.signalStrength = attachDeviceAll[key]["Signal"] + "%";
                        //}
                        //else
                        //{
                        //    NodeDevice.signalStrength = "%";
                        //}

                        //NodeDevice.MACaddress = key;
                        //NodeDevice.connectType = attachDeviceAll[key]["Connect"];
                        //if (attachDeviceAll[key].ContainsKey("AccessControl"))
                        //{
                        //    NodeDevice.AccessControl = attachDeviceAll[key]["AccessControl"];
                        //}
                        //else
                        //{
                        //    NodeDevice.AccessControl = "";
                        //}
                        NodeDevice.IPaddress = deviceInfo.ElementAt(0);
                        if (deviceInfo.Count > 4)
                        {
                            NodeDevice.linkRate = deviceInfo.ElementAt(4) + "Mbps";
                        }
                        else
                        {
                            NodeDevice.linkRate = "Mbps";
                        }

                        if (deviceInfo.Count > 5)
                        {
                            NodeDevice.signalStrength = deviceInfo.ElementAt(5) + "%";
                        }
                        else
                        {
                            NodeDevice.signalStrength = "%";
                        }

                        NodeDevice.MACaddress = deviceInfo.ElementAt(2);
                        NodeDevice.connectType = deviceInfo.ElementAt(3);
                        if (deviceInfo.Count > 6)
                        {
                            NodeDevice.AccessControl = deviceInfo.ElementAt(6);
                        }
                        else
                        {
                            NodeDevice.AccessControl = "";
                        }
                        //var group = new DeviceGroup(NodeDevice);
                        var group = new DeviceGroup(NodeDevice);
                        this.DeviceGroups.Add(group);
                    }
                }   //foreach

                int num = 0;
                //foreach (string key in attachDeviceAll.Keys)
                foreach (List<string> deviceInfo in attachDeviceAll)
                {
                    num++;
                    Node NodeDevice = new Node();

                    //if (loacalIp != attachDeviceAll[key]["Ip"])
                    if (loacalIp != deviceInfo.ElementAt(0))
                    {
                        NodeDevice.uniqueId = "Device" + num.ToString();
                        bool bFound = false;
                        if (NetworkMapInfo.fileContent != "")
                        {
                            string[] AllDeviceInfo = NetworkMapInfo.fileContent.Split(';');
                            for (int i = 0; i < AllDeviceInfo.Length; i++)
                            {
                                if (AllDeviceInfo[i] != "" && AllDeviceInfo[i] != null)
                                {
                                    string[] DeviceInfo = AllDeviceInfo[i].Split(',');
                                    //if (DeviceInfo[0] == key)
                                    if (DeviceInfo[0] == deviceInfo.ElementAt(2))
                                    {
                                        bFound = true;
                                        if (DeviceInfo[1] != "")
                                            NodeDevice.deviceName = DeviceInfo[1];
                                        else
                                            //NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                                            NodeDevice.deviceName = deviceInfo.ElementAt(1);

                                        if (DeviceInfo[2] != "")
                                            NodeDevice.deviceType = DeviceInfo[2];
                                        else
                                            NodeDevice.deviceType = "networkdev";
                                    }
                                    //else
                                    //{
                                    //    NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                                    //    NodeDevice.deviceType = "networkdev";
                                    //}
                                }
                            }
                            if (!bFound)
                            {
                                //NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                                NodeDevice.deviceName = deviceInfo.ElementAt(1);
                                NodeDevice.deviceType = "networkdev";
                            }
                        }
                        else
                        {
                            //NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                            NodeDevice.deviceName = deviceInfo.ElementAt(1);
                            NodeDevice.deviceType = "networkdev";
                        }
                        //NodeDevice.deviceName = attachDeviceAll[key]["HostName"];
                        //NodeDevice.deviceType = attachDeviceAll[key]["Connect"];
                        //NodeDevice.IPaddress = attachDeviceAll[key]["Ip"];
                        //if (attachDeviceAll[key].ContainsKey("LinkSpeed"))
                        //{
                        //    NodeDevice.linkRate = attachDeviceAll[key]["LinkSpeed"] + "Mbps";
                        //}
                        //else
                        //{
                        //    NodeDevice.linkRate = "Mbps";
                        //}

                        //if (attachDeviceAll[key].ContainsKey("Signal"))
                        //{
                        //    NodeDevice.signalStrength = attachDeviceAll[key]["Signal"] + "%";
                        //}
                        //else
                        //{
                        //    NodeDevice.signalStrength = "%";
                        //}
                        //NodeDevice.MACaddress = key;
                        //NodeDevice.connectType = attachDeviceAll[key]["Connect"];
                        //if (attachDeviceAll[key].ContainsKey("AccessControl"))
                        //{
                        //    NodeDevice.AccessControl = attachDeviceAll[key]["AccessControl"];
                        //}
                        //else
                        //{
                        //    NodeDevice.AccessControl = "";
                        //}
                        NodeDevice.IPaddress = deviceInfo.ElementAt(0);
                        if (deviceInfo.Count > 4)
                        {
                            NodeDevice.linkRate = deviceInfo.ElementAt(4) + "Mbps";
                        }
                        else
                        {
                            NodeDevice.linkRate = "Mbps";
                        }

                        if (deviceInfo.Count > 5)
                        {
                            NodeDevice.signalStrength = deviceInfo.ElementAt(5) + "%";
                        }
                        else
                        {
                            NodeDevice.signalStrength = "%";
                        }
                        NodeDevice.MACaddress = deviceInfo.ElementAt(2);
                        NodeDevice.connectType = deviceInfo.ElementAt(3);
                        if (deviceInfo.Count > 6)
                        {
                            NodeDevice.AccessControl = deviceInfo.ElementAt(6);
                        }
                        else
                        {
                            NodeDevice.AccessControl = "";
                        }
                        var group = new DeviceGroup(NodeDevice);
                        this.DeviceGroups.Add(group);
                    }
                }   //foreach
            }   //else
        }
    }
}
